extension/src/minn_maxn.rs

```diff
@@ -0,0 +1,1159 @@
+    }
+}
+
+impl<T: Ord + Copy> From<&[T]> for NMostTransState<T> {
+    fn from(vals: &[T]) -> Self {
+        let mut state = Self::new(vals.len(), vals[0]);
+        for val in vals[1..].iter() {
+            state.new_entry(*val);
+        }
+        state
+    }
+}
+
+fn min_max_trans_function<T: Ord>(
+    state: Option<Inner<NMostTransState<T>>>,
+    val: T,
+    capacity: usize,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<NMostTransState<T>>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            if state.is_none() {
+                return Internal::new(NMostTransState::<T>::new(capacity, val)).to_inner();
+            }
+
+            let mut state = state.unwrap();
+            state.new_entry(val);
+            Some(state)
+        })
+    }
+}
+
+fn min_max_rollup_trans_function<T: Ord + Copy>(
+    state: Option<Inner<NMostTransState<T>>>,
+    sorted_vals: &[T],
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<NMostTransState<T>>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            if let Some(mut state) = state {
+                for val in sorted_vals {
+                    // The values are sorted, so as soon as we find one that shouldn't be added, we're done
+                    if !state.belongs_in_heap(val) {
+                        return Some(state);
+                    }
+                    state.new_entry(*val);
+                }
+
+                Some(state)
+            } else {
+                Internal::new::<NMostTransState<T>>(sorted_vals.into()).to_inner()
+            }
+        })
+    }
+}
+
+fn min_max_trans_combine<T: Clone + Ord + Copy>(
+    first: Option<Inner<NMostTransState<T>>>,
+    second: Option<Inner<NMostTransState<T>>>,
+) -> Option<Inner<NMostTransState<T>>> {
+    match (first, second) {
+        (None, None) => None,
+        (None, Some(only)) | (Some(only), None) => unsafe {
+            Internal::new(only.clone()).to_inner()
+        },
+        (Some(a), Some(b)) => {
+            let mut a = a.clone();
+            // This could be made more efficient by iterating in the appropriate order with an early exit, but would requiring ordering the other heap
+            for entry in b.heap.iter() {
+                a.new_entry(*entry);
+            }
+            unsafe { Internal::new(a).to_inner() }
+        }
+    }
+}
+
+// Note that min_n queries store a max heap of the lowest values, which is the normal behavior of Rust's BinaryHeap
+// Max_n queries on the other hand, store a min heap of the highest values, which requires a Reverse<'type'> as the type of BinaryHeap.
+type IntMinTransType = NMostTransState<i64>;
+type IntMaxTransType = NMostTransState<Reverse<i64>>;
+type FloatMinTransType = NMostTransState<NotNan<f64>>;
+type FloatMaxTransType = NMostTransState<Reverse<NotNan<f64>>>;
+type TimeMinTransType = NMostTransState<pg_sys::TimestampTz>;
+type TimeMaxTransType = NMostTransState<Reverse<pg_sys::TimestampTz>>;
```

Comment:
What do you think about keeping this file for the common type and functions and creating one child module file per type?  I think it would not only make this file and each individual implementation easier to read, but if kept in the same order, it makes it easy to diff the implementations to spot interesting differences.

Reply:
Can't say I like it, but I think I dislike it less than any of the alternatives.